File: app/src/main/java/com/example/gastosapp/Fragments/FragmentPresupuesto.java
package com.example.gastosapp.Fragments;

import android.os.Bundle;
import android.os.Handler;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;

import com.airbnb.lottie.LottieAnimationView;
import com.example.gastosapp.Models.Presupuesto;
import com.example.gastosapp.R;
import com.example.gastosapp.viewModels.PresupuestoViewModel;

import java.util.List;

public class FragmentPresupuesto extends Fragment {

    private static final String ARG_PARAM1 = "param1";
    private static final String ARG_PARAM2 = "param2";
    private String mParam1;
    private String mParam2;

    // VARIABLE CORREGIDA: PresupuestoViewModel (con P mayúscula)
    private PresupuestoViewModel viewModel;
    private LinearLayout containerPresupuestos;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // ✅ ViewModel a nivel de Activity para persistencia entre Fragments
        viewModel = new ViewModelProvider(requireActivity()).get(PresupuestoViewModel.class);

        // ✅ INICIALIZAR CON PERSISTENCIA
        viewModel.init(requireContext());

        System.out.println("🎯 FragmentPresupuesto creado");
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_presupuesto, container, false);

        containerPresupuestos = view.findViewById(R.id.containerPresupuestos);
        LottieAnimationView btnAddCategory = view.findViewById(R.id.agregarPresupuesto);

        System.out.println("🔍 Vistas inicializadas");

        // Observer para los presupuestos
        viewModel.getPresupuestos().observe(getViewLifecycleOwner(), new Observer<List<Presupuesto>>() {
            @Override
            public void onChanged(List<Presupuesto> presupuestos) {
                System.out.println("👀 Datos actualizados: " + presupuestos.size() + " presupuestos");
                actualizarVistaPresupuestos(presupuestos);
            }
        });

        btnAddCategory.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                System.out.println("🎯 Botón agregar presionado");
                btnAddCategory.playAnimation();

                new Handler().postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        showFloatingWindow();
                    }
                }, 500);
            }
        });

        return view;
    }

    private void actualizarVistaPresupuestos(List<Presupuesto> presupuestos) {
        System.out.println("Actualizando vista con " + presupuestos.size() + " presupuestos");

        // Limpiar el contenedor
        containerPresupuestos.removeAllViews();

        if (presupuestos.isEmpty()) {
            // Mostrar estado vacío
            TextView tvEmpty = new TextView(requireContext());
            tvEmpty.setText("No hay presupuestos. ¡Agrega uno nuevo! ");
            tvEmpty.setTextSize(16);
            tvEmpty.setGravity(View.TEXT_ALIGNMENT_CENTER);
            tvEmpty.setPadding(0, 50, 0, 50);
            tvEmpty.setTextColor(getResources().getColor(android.R.color.darker_gray));
            containerPresupuestos.addView(tvEmpty);

            System.out.println("Mostrando estado vacío");
        } else {
            // Agregar cada presupuesto a la vista
            for (int i = 0; i < presupuestos.size(); i++) {
                Presupuesto presupuesto = presupuestos.get(i);
                View itemView = crearItemPresupuesto(presupuesto, i);
                containerPresupuestos.addView(itemView);
            }
            System.out.println(" " + presupuestos.size() + " presupuestos mostrados");
        }
    }

    private void showFloatingWindow() {
        System.out.println("Mostrando diálogo de agregar presupuesto");

        try {
            FragmentAgregarPresupuesto dialogFragment = new FragmentAgregarPresupuesto();

            // Configurar el listener para cuando se guarde un presupuesto
            dialogFragment.setPresupuestoGuardadoListener(new FragmentAgregarPresupuesto.PresupuestoGuardadoListener() {
                @Override
                public void onPresupuestoGuardado(Presupuesto presupuesto) {
                    System.out.println("Presupuesto guardado recibido: " + presupuesto.getNombre());

                    // Agregar el presupuesto al ViewModel
                    viewModel.agregarPresupuesto(presupuesto);

                    Toast.makeText(requireContext(), "¡Presupuesto agregado!", Toast.LENGTH_SHORT).show();
                }
            });

            // Mostrar el diálogo
            dialogFragment.show(getParentFragmentManager(), "presupuesto_dialog");

        } catch (Exception e) {
            System.out.println("Error al mostrar diálogo: " + e.getMessage());
            Toast.makeText(requireContext(), "Error al abrir formulario", Toast.LENGTH_SHORT).show();
        }
    }

    private View crearItemPresupuesto(Presupuesto presupuesto, int position) {
        System.out.println("Creando item para: " + presupuesto.getNombre());

        LayoutInflater inflater = LayoutInflater.from(requireContext());
        View itemView = inflater.inflate(R.layout.item_presupuesto, containerPresupuestos, false);

        try {
            // Configurar las vistas del item
            TextView tvNombre = itemView.findViewById(R.id.tvNombrePresupuesto);
            TextView tvCantidad = itemView.findViewById(R.id.tvCantidad);
            TextView tvFechaInicio = itemView.findViewById(R.id.tvFechaInicio);
            TextView tvFechaFinal = itemView.findViewById(R.id.tvFechaFinal);
            TextView tvEstado = itemView.findViewById(R.id.tvEstado);

            // Establecer los datos
            tvNombre.setText(presupuesto.getNombre());
            tvCantidad.setText(String.format("$%.2f", presupuesto.getCantidad()));
            tvFechaInicio.setText(presupuesto.getFechaInicio());
            tvFechaFinal.setText(presupuesto.getFechaFinal());
            tvEstado.setText("Activo");

            // Configurar botón de eliminar
            View btnEliminar = itemView.findViewById(R.id.btnEliminar);
            if (btnEliminar != null) {
                btnEliminar.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        eliminarPresupuesto(position);
                    }
                });
            }

        } catch (Exception e) {
            System.out.println("Error al configurar item: " + e.getMessage());
        }

        return itemView;
    }

    private void eliminarPresupuesto(int position) {
        System.out.println("🗑Eliminando presupuesto en posición: " + position);

        // AHORA SÍ EXISTE este método en el ViewModel
        viewModel.eliminarPresupuesto(position);
        Toast.makeText(requireContext(), "Presupuesto eliminado", Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onResume() {
        super.onResume();
        System.out.println("FragmentPresupuesto: onResume");
    }

    @Override
    public void onPause() {
        super.onPause();
        System.out.println("FragmentPresupuesto: onPause");
    }
}

File: app/src/main/java/com/example/gastosapp/Fragments/FragmentPresupuesto.kt
package com.example.gastosapp.Fragments

import android.os.Bundle
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import android.widget.Toast
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import com.example.gastosapp.Models.Categoria
import com.example.gastosapp.Models.Presupuesto
import com.example.gastosapp.R
import com.example.gastosapp.databinding.FragmentPresupuestoBinding
import com.example.gastosapp.viewModels.PresupuestoViewModel
import java.text.SimpleDateFormat
import java.util.*

class FragmentPresupuesto : Fragment() {

    private var _binding: FragmentPresupuestoBinding? = null
    private val binding get() = _binding!!

    private val vm: PresupuestoViewModel by activityViewModels()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        _binding = FragmentPresupuestoBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        vm.presupuestos.observe(viewLifecycleOwner) { actualizarUI(it) }

        binding.agregarPresupuesto.setOnClickListener {
            binding.agregarPresupuesto.playAnimation()
            abrirAgregar()
        }
    }

    private fun abrirAgregar() {
        FragmentAgregarPresupuesto().apply {
            arguments = Bundle().apply {
                putStringArrayList("categorias_ocupadas", ArrayList(obtenerCategoriasOcupadas()))
            }
            setOnPresupuestoSaved { vm.agregarPresupuesto(it) }
        }.show(parentFragmentManager, "agregar_presupuesto")
    }

    private fun abrirEdicion(p: Presupuesto) {
        FragmentAgregarPresupuesto().apply {
            arguments = Bundle().apply {
                putStringArrayList("categorias_ocupadas", ArrayList(obtenerCategoriasOcupadas(p.id)))
            }
            editarPresupuesto(p) { vm.editarPresupuesto(it) }
        }.show(parentFragmentManager, "editar_presupuesto")
    }

    private fun obtenerCategoriasOcupadas(excluirId: String? = null): List<String> {
        val hoy = Calendar.getInstance().apply {
            set(Calendar.HOUR_OF_DAY, 0); set(Calendar.MINUTE, 0); set(Calendar.SECOND, 0); set(Calendar.MILLISECOND, 0)
        }

        return vm.presupuestos.value?.filter { p ->
            p.id != excluirId && p.fechaFinal.after(hoy.time) || sameDay(p.fechaFinal, hoy.time)
        }?.map { it.categoriaNombre } ?: emptyList()
    }

    private fun actualizarUI(lista: List<Presupuesto>) {
        binding.containerPresupuestos.removeAllViews()

        if (lista.isEmpty()) {
            binding.containerPresupuestos.addView(TextView(requireContext()).apply {
                text = "No hay presupuestos. ¡Agrega uno!"
                textSize = 18f
                gravity = Gravity.CENTER
                setPadding(0, 200, 0, 200)
                setTextColor(ContextCompat.getColor(requireContext(), android.R.color.darker_gray))
            })
            return
        }

        lista.forEach { p ->
            binding.containerPresupuestos.addView(crearItem(p))
        }
    }

    private fun crearItem(p: Presupuesto): View {
        val v = layoutInflater.inflate(R.layout.item_presupuesto, binding.containerPresupuestos, false)

        v.findViewById<TextView>(R.id.tvNombrePresupuesto).text = p.categoriaNombre
        v.findViewById<TextView>(R.id.tvCantidad).text = String.format("$%.2f", p.cantidad)
        v.findViewById<TextView>(R.id.tvFechaInicio).text = formatearFecha(p.fechaInicio)
        v.findViewById<TextView>(R.id.tvFechaFinal).text = formatearFecha(p.fechaFinal)
        v.findViewById<TextView>(R.id.tvCategoria).text = p.categoriaNombre

        val saldo = p.cantidad - p.montoGastado
        val tvSaldo = v.findViewById<TextView>(R.id.tvSaldoDisponible)
        tvSaldo.text = String.format("Saldo: $%.2f", saldo)
        tvSaldo.setTextColor(when {
            saldo < 0 -> ContextCompat.getColor(requireContext(), android.R.color.holo_red_dark)
            saldo < p.cantidad * 0.2 -> ContextCompat.getColor(requireContext(), android.R.color.holo_orange_dark)
            else -> ContextCompat.getColor(requireContext(), android.R.color.holo_green_dark)
        })

        val tvEstado = v.findViewById<TextView>(R.id.tvEstadoPresupuesto)
        val expirado = p.fechaFinal.before(Date())
        tvEstado.text = if (expirado) "Expirado" else "Activo"
        tvEstado.setBackgroundResource(if (expirado) R.drawable.bg_estado_expirado else R.drawable.bg_estado_activo)

        v.findViewById<View>(R.id.btnEditar).setOnClickListener { abrirEdicion(p) }
        v.findViewById<View>(R.id.btnEliminar).setOnClickListener {
            vm.eliminarPresupuesto(p)
            Toast.makeText(requireContext(), "Presupuesto eliminado", Toast.LENGTH_SHORT).show()
        }

        return v
    }

    private fun sameDay(d1: Date, d2: Date): Boolean {
        val fmt = SimpleDateFormat("yyyyMMdd", Locale.getDefault())
        return fmt.format(d1) == fmt.format(d2)
    }

    private fun formatearFecha(date: Date): String {
        return SimpleDateFormat("dd/MM/yyyy", Locale.getDefault()).format(date)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}